tools/model.go

```diff
@@ -0,0 +1,286 @@
+				Visulizations:     nil,
+				ObservationalData: nil,
+			},
+		},
+	}
+
+	for _, p := range rm.Metadata.PlanFiles {
+		mod.Files.InputFiles.ControlFiles.Paths = append(mod.Files.InputFiles.ControlFiles.Paths, p.Path)
+		mod.Files.InputFiles.ControlFiles.Data["PlanTitle"] = p.PlanTitle
+	}
+	for _, g := range rm.Metadata.GeomFiles {
+		mod.Files.InputFiles.GeometryFiles.Paths = append(mod.Files.InputFiles.GeometryFiles.Paths, g.Path)
+		mod.Files.InputFiles.GeometryFiles.FeaturesProperties["GeomTitle"] = g.GeomTitle
+	}
+	for _, f := range rm.Metadata.FlowFiles {
+		mod.Files.InputFiles.ForcingFiles.Paths = append(mod.Files.InputFiles.ForcingFiles.Paths, f.Path)
+		mod.Files.InputFiles.ForcingFiles.Data["FlowTitle"] = f.FlowTitle
+	}
+
+	// need to add output files and supplemental files...
+
+	return mod, nil
+}
+
+// GeospatialData ...
+func (rm *RasModel) GeospatialData() (GeoData, error) {
+	gd := GeoData{Features: make(map[string]Features), Georeference: DestinationCRS}
+
+	for _, g := range rm.Metadata.GeomFiles {
+		if err := GetGeospatialData(&gd, rm.FileStore, g.Path); err != nil {
+			return gd, err
+		}
+	}
+	return gd, nil
+}
+
+func getModelFiles(rm *RasModel) error {
+	prefix := filepath.Dir(rm.Metadata.ProjFilePath) + "/"
+
+	files, err := rm.FileStore.GetDir(prefix, false)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range *files {
+		if filepath.Ext(file.Name) != ".prj" {
+			rm.FileList = append(rm.FileList, filepath.Join(file.Path, file.Name))
+		}
+	}
+
+	return nil
+}
+
+func NewRasModel(key string, fs filestore.FileStore) (*RasModel, error) {
+	rm := RasModel{ModelDirectory: filepath.Dir(key), FileStore: fs, Type: "RAS"}
+
+	err := verifyPrjPath(key, &rm)
+	if err != nil {
+		return &rm, err
+	}
+
+	err = getModelFiles(&rm)
+	if err != nil {
+		return &rm, err
+	}
+
+	err = getPrjData(&rm)
+	if err != nil {
+		return &rm, err
+	}
+
+	errChan := make(chan error)
+	var rasWG rasWaitGroup
+
+	for _, fp := range rm.FileList {
+
+		ext := filepath.Ext(fp)
+
+		switch {
+
+		case rasRE.Plan.MatchString(ext):
+			rasWG.Plan.Add(1)
+			go getPlanData(&rm, fp, &rasWG.Plan, errChan)
+
+		case rasRE.Geom.MatchString(ext):
+			rasWG.Geom.Add(1)
+			go getGeomData(&rm, fp, &rasWG.Geom, errChan)
+
+		case rasRE.AllFlow.MatchString(ext):
+			rasWG.Flow.Add(1)
+			go getFlowData(&rm, fp, &rasWG.Flow, errChan)
+
+		}
+	}
+
+	rasWG.Plan.Wait()
+	rasWG.Geom.Wait()
+	rasWG.Flow.Wait()
+
+	if len(errChan) > 0 {
+		fmt.Printf("Encountered %d errors\n", len(errChan))
+		return &rm, <-errChan
+	}
+
+	for _, p := range rm.Metadata.PlanFiles {
+		rm.Version += fmt.Sprintf("%s: %s, ", p.FileExt, p.ProgramVersion)
+	}
+	for _, g := range rm.Metadata.GeomFiles {
+		rm.Version += fmt.Sprintf("%s: %s, ", g.FileExt, g.ProgramVersion)
+	}
+	for _, f := range rm.Metadata.FlowFiles {
+		rm.Version += fmt.Sprintf("%s: %s, ", f.FileExt, f.ProgramVersion)
+	}
+
+	rm.Version = rm.Version[0 : len(rm.Version)-2]
```

Comment:
We've seen the code get to here an panic because the version has zero length. Could a short if statement could be added here to avoid the panic and set the rm.isModel to false?

Reply:
I checked a few models and some do not have the version even though they have the .prj, .g, .f, and .p files. I added an if statement to check the length of the version string before it is sliced so that it does not panic, but I did not set rm.IsModel to false.